File: UIKit and Animations/WWDC 2014/A Look Inside Presentation Controllers (With a case study on alerts and search)/Sample_Codes/LookInside Presentation Controllers, Adaptivity, and Custom Animator Objects/LookInsidePresentationControllersAdaptivityandCustomAnimatorObjects/LookInside/AAPLCoolPresentationController.m
/*
 Copyright (C) 2014 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 
  AAPLCoolPresentationController implementation.
  
 */

#import "AAPLCoolPresentationController.h"

@implementation AAPLCoolPresentationController

- (instancetype)initWithPresentingViewController:(UIViewController *)presentingViewController presentedViewController:(UIViewController *)presentedViewController
{
    self = [super initWithPresentedViewController:presentedViewController presentingViewController:presentingViewController];
    if(self)
    {
        dimmingView = [[UIView alloc] init];
        [dimmingView setBackgroundColor:[[UIColor purpleColor] colorWithAlphaComponent:0.4]];
        
        bigFlowerImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"BigFlower"]];
        carlImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"Carl"]];
        [carlImageView setFrame:CGRectMake(0,0,500,245)];
        
        jaguarPrintImageH = [[UIImage imageNamed:@"JaguarH"] resizableImageWithCapInsets:UIEdgeInsetsZero resizingMode:UIImageResizingModeTile];
        jaguarPrintImageV = [[UIImage imageNamed:@"JaguarV"] resizableImageWithCapInsets:UIEdgeInsetsZero resizingMode:UIImageResizingModeTile];

        topJaguarPrintImageView = [[UIImageView alloc] initWithImage:jaguarPrintImageH];
        bottomJaguarPrintImageView = [[UIImageView alloc] initWithImage:jaguarPrintImageH];

        leftJaguarPrintImageView = [[UIImageView alloc] initWithImage:jaguarPrintImageV];
        rightJaguarPrintImageView = [[UIImageView alloc] initWithImage:jaguarPrintImageV];
    }
    return self;
}

- (CGRect)frameOfPresentedViewInContainerView
{
    CGRect containerBounds = [[self containerView] bounds];
    
    CGRect presentedViewFrame = CGRectZero;
    presentedViewFrame.size = CGSizeMake(300, 500);
    presentedViewFrame.origin = CGPointMake(containerBounds.size.width / 2.0, containerBounds.size.height / 2.0);
    presentedViewFrame.origin.x -= presentedViewFrame.size.width / 2.0;
    presentedViewFrame.origin.y -= presentedViewFrame.size.height / 2.0;
    
    return presentedViewFrame;
}

- (void)presentationTransitionWillBegin
{
    [super presentationTransitionWillBegin];
    
    [self addViewsToDimmingView];

    [dimmingView setAlpha:0.0];
    
    [[[self presentedViewController] transitionCoordinator] animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext> context) {
		[self->dimmingView setAlpha:1.0];
    } completion:nil];

    [self moveJaguarPrintToPresentedPosition:NO];
    
    [UIView animateWithDuration:1.0 animations:^{
        [self moveJaguarPrintToPresentedPosition:YES];
    }];
}

- (void)containerViewWillLayoutSubviews
{
    [dimmingView setFrame:[[self containerView] bounds]];
}

- (void)containerViewDidLayoutSubviews
{
    CGPoint bigFlowerCenter = [dimmingView frame].origin;
    bigFlowerCenter.x += [[bigFlowerImageView image] size].width / 4.0;
    bigFlowerCenter.y += [[bigFlowerImageView image] size].height / 4.0;
    
    [bigFlowerImageView setCenter:bigFlowerCenter];
    
    CGRect carlFrame = [carlImageView frame];
    carlFrame.origin.y = [dimmingView bounds].size.height - carlFrame.size.height;
    
    [carlImageView setFrame:carlFrame];
}

- (void)dismissalTransitionWillBegin
{
    [super dismissalTransitionWillBegin];

    [[[self presentedViewController] transitionCoordinator] animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext> context) {
        [dimmingView setAlpha:0.0];
} completion:nil];
}

- (void)addViewsToDimmingView
{
    [dimmingView addSubview:bigFlowerImageView];
    [dimmingView addSubview:carlImageView];

    [dimmingView addSubview:topJaguarPrintImageView];
    [dimmingView addSubview:bottomJaguarPrintImageView];

    [dimmingView addSubview:leftJaguarPrintImageView];
    [dimmingView addSubview:rightJaguarPrintImageView];
    
    [[self containerView] addSubview:dimmingView];
}

- (void)moveJaguarPrintToPresentedPosition:(BOOL)presentedPosition
{
    CGSize horizontalJaguarSize = [jaguarPrintImageH size];
    CGSize verticalJaguarSize = [jaguarPrintImageV size];
    CGRect frameOfView = [self frameOfPresentedViewInContainerView];
    CGRect containerFrame = [[self containerView] frame];

    CGRect topFrame, bottomFrame, leftFrame, rightFrame;
    topFrame.size.height = bottomFrame.size.height = horizontalJaguarSize.height;
    topFrame.size.width = bottomFrame.size.width = frameOfView.size.width;

    leftFrame.size.width = rightFrame.size.width = verticalJaguarSize.width;
    leftFrame.size.height = rightFrame.size.height = frameOfView.size.height;

    topFrame.origin.x = frameOfView.origin.x;
    bottomFrame.origin.x = frameOfView.origin.x;

    leftFrame.origin.y = frameOfView.origin.y;
    rightFrame.origin.y = frameOfView.origin.y;

    CGRect frameToAlignAround = presentedPosition ? frameOfView : containerFrame;

    topFrame.origin.y = CGRectGetMinY(frameToAlignAround) - horizontalJaguarSize.height;
    bottomFrame.origin.y = CGRectGetMaxY(frameToAlignAround);
    leftFrame.origin.x = CGRectGetMinX(frameToAlignAround) - verticalJaguarSize.width;
    rightFrame.origin.x = CGRectGetMaxX(frameToAlignAround);
    
    [topJaguarPrintImageView setFrame:topFrame];
    [bottomJaguarPrintImageView setFrame:bottomFrame];
    [leftJaguarPrintImageView setFrame:leftFrame];
    [rightJaguarPrintImageView setFrame:rightFrame];
}

@end

File: UIKit and Animations/WWDC 2014/A Look Inside Presentation Controllers (With a case study on alerts and search)/Sample_Codes/LookInside Presentation Controllers, Adaptivity, and Custom Animator Objects/LookInside_Swift/LookInside_Swift/AAPLCoolPresentationController.swift
//
//  AAPLCoolPresentationController.swift
//  LookInside_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/18/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

class AAPLCoolPresentationController: UIPresentationController {
	
	var bigFlowerImageView: UIImageView!
	var carlImageView: UIImageView!
	
	var jaguarPrintImageH: UIImage!
	var jaguarPrintImageV: UIImage!

	var topJaguarPrintImageView: UIImageView!
	var bottomJaguarPrintImageView: UIImageView!
	
	var leftJaguarPrintImageView: UIImageView!
	var rightJaguarPrintImageView: UIImageView!

	var dimmingView: UIView!
	
	override init(presentedViewController: UIViewController, presenting presentingViewController: UIViewController?) {
		super.init(presentedViewController: presentedViewController, presenting: presentingViewController)
		
		self.dimmingView = UIView()
		self.dimmingView.backgroundColor = UIColor.purple.withAlphaComponent(0.4)
		self.bigFlowerImageView = UIImageView(image: UIImage(named: "BigFlower"))
		self.carlImageView = UIImageView(image: UIImage(named: "Carl"))
		self.carlImageView.frame = CGRect(x: 0, y: 0, width: 500, height: 245)
		
		self.jaguarPrintImageH = UIImage(named: "JaguarH")?.resizableImage(withCapInsets: .zero, resizingMode: .tile)
		self.jaguarPrintImageV = UIImage(named: "JaguarV")?.resizableImage(withCapInsets: .zero, resizingMode: .tile)
		
		self.topJaguarPrintImageView = UIImageView(image: jaguarPrintImageH)
		self.bottomJaguarPrintImageView = UIImageView(image: jaguarPrintImageH)
		
		self.leftJaguarPrintImageView = UIImageView(image: jaguarPrintImageV)
		self.rightJaguarPrintImageView = UIImageView(image: jaguarPrintImageV)
	}
	
	override var frameOfPresentedViewInContainerView: CGRect {
		let containerBounds = self.containerView!.bounds
		var presentedViewFrame = CGRect.zero
		presentedViewFrame.size = CGSize(width: 300, height: 500)
		presentedViewFrame.origin = CGPoint(x: containerBounds.size.width / 2.0, y: containerBounds.size.height / 2.0)
		presentedViewFrame.origin.x -= presentedViewFrame.size.width / 2.0
		presentedViewFrame.origin.y -= presentedViewFrame.size.height / 2.0
		return presentedViewFrame
	}
	
	override func presentationTransitionWillBegin() {
		super.presentationTransitionWillBegin()
		
		self.addViewsToDimmingView()
		
		self.dimmingView.alpha = 0.0
		
		self.presentedViewController.transitionCoordinator?.animate(alongsideTransition: { (context) in
			self.dimmingView.alpha = 1.0
		}, completion: nil)

		self.moveJaguarPrintToPresentedPosition(false)
		
		UIView.animate(withDuration: 1.0) {
			self.moveJaguarPrintToPresentedPosition(true)
		}
	}
	
	override func containerViewWillLayoutSubviews() {
		self.dimmingView.frame = self.containerView!.bounds
	}
	
	override func containerViewDidLayoutSubviews() {
		var bigFlowerCenter = self.dimmingView.frame.origin
		bigFlowerCenter.x += self.bigFlowerImageView.image!.size.width/4.0
		bigFlowerCenter.y += self.bigFlowerImageView.image!.size.height/4.0
		
		self.bigFlowerImageView.center = bigFlowerCenter
		
		var carlFrame = self.carlImageView.frame
		carlFrame.origin.y = self.dimmingView.bounds.size.height - carlFrame.size.height
		
		self.carlImageView.frame = carlFrame
	}
	
	override func dismissalTransitionWillBegin() {
		super.dismissalTransitionWillBegin()
		
		self.presentedViewController.transitionCoordinator?.animate(alongsideTransition: { (context) in
			self.dimmingView.alpha = 0.0
		}, completion: nil)
	}
	
	func addViewsToDimmingView() {
		self.dimmingView.addSubview(bigFlowerImageView)
		self.dimmingView.addSubview(carlImageView)

		self.dimmingView.addSubview(topJaguarPrintImageView)
		self.dimmingView.addSubview(bottomJaguarPrintImageView)

		self.dimmingView.addSubview(leftJaguarPrintImageView)
		self.dimmingView.addSubview(rightJaguarPrintImageView)

		self.containerView?.addSubview(self.dimmingView)
	}
	
	func moveJaguarPrintToPresentedPosition(_ presentedPosition: Bool) {
		let horizontalJaguarSize = self.jaguarPrintImageH.size
		let verticalJaguarSize = self.jaguarPrintImageV.size
		let frameOfView = self.frameOfPresentedViewInContainerView
		let containerFrame =  self.containerView?.frame
		
		var topFrame = CGRect.zero
		var bottomFrame = CGRect.zero
		var leftFrame = CGRect.zero
		var rightFrame = CGRect.zero

		topFrame.size.height = horizontalJaguarSize.height
		topFrame.size.width = frameOfView.size.width
		bottomFrame.size.height = topFrame.size.height
		bottomFrame.size.width = topFrame.size.width

		leftFrame.size.width = verticalJaguarSize.width
		leftFrame.size.height = frameOfView.size.height
		rightFrame.size.width = leftFrame.size.width
		rightFrame.size.height = leftFrame.size.height

		topFrame.origin.x = frameOfView.origin.x
		bottomFrame.origin.x = frameOfView.origin.x
		
		leftFrame.origin.y = frameOfView.origin.y
		rightFrame.origin.y = frameOfView.origin.y
		
		let frameToAlignAround = presentedPosition ? frameOfView : containerFrame
		
		topFrame.origin.y = frameToAlignAround!.minY - horizontalJaguarSize.height;
		bottomFrame.origin.y = frameToAlignAround!.maxY
		leftFrame.origin.x = frameToAlignAround!.minX - verticalJaguarSize.width;
		rightFrame.origin.x = frameToAlignAround!.maxX
		
		self.topJaguarPrintImageView.frame = topFrame
		self.bottomJaguarPrintImageView.frame = bottomFrame
		self.leftJaguarPrintImageView.frame = leftFrame
		self.rightJaguarPrintImageView.frame = rightFrame
	}
}
